lib/Controller.php

```diff
@@ -151,6 +151,8 @@ public function __construct()
             header('Access-Control-Allow-Origin: *');
             header('Access-Control-Allow-Methods: GET, POST, PUT, DELETE');
             header('Access-Control-Allow-Headers: X-Requested-With, Content-Type');
+            header('X-Uncompressed-Content-Length: ' . strlen($this->_json));
```

Comment:
No issue IMHO with adding this extra information, just one question: Why could we not set the more standard `Content-Length` header?

Is it to deal with webservers that do compression and then replace this header with the compressed size? I assume you'd probably want to compressed size for the loading bar, when content got compressed. The bar should reach the next stage (decryption) when either the compressed or uncompressed length is reached, but if the header indicates too large a value it will "jump forward" once the smaller compressed size is transmitted.

Reply:
> Is it to deal with webservers that do compression and then replace this header with the compressed size?

This is indeed the exact conclusion I reached :slightly_smiling_face:

I tried to set `Content-Length`, but it didn't make it through (it didn't appear on the response received client-side).

If I understand correctly, web servers set `Content-Length` when they know the total size of the data about to be transmitted. And they know it when the data comes from a file (requesting `index.html` for instance), but they can't know it when it comes from PHP output, which is our case here.

When web servers set `Content-Length` for a compressed response, it is indeed set to the size of the compressed data.

However, when `Content-Length` is not set, then `XMLHttpRequest`'s `ProgressEvent.loaded`, in Chrome, returns the size of _decompressed_ data. Which is great for us, as we can simply do the ratio with the value in our custom header `X-Uncompressed-Content-Length`. (In Firefox, it's different, which is why I created an [issue in the spec](https://github.com/whatwg/xhr/issues/388).)

If you want to see a working example of this, have a look at https://babel.zwyx.dev/ (have a quick read about the concept, find a book, click Share, and use the access link :wink:).

Reply:
Ok, that convinces me to give this a try as is. We can always update on it later on.